File: backend/src/nodeServer.ts
import express from 'express';
import multer from 'multer';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const app = express();
const NODE_NAME = process.env.NODE_NAME || "node1";
const STORAGE_DIR = path.join("storage", NODE_NAME);

if (!fs.existsSync(STORAGE_DIR)) {
    fs.mkdirSync(STORAGE_DIR, { recursive: true });
}

// Multer disk storage config
const storage = multer.diskStorage({
    destination: (req, file, cb) => cb(null, STORAGE_DIR),
    filename: (req, file, cb) => cb(null, req.params.filename || file.originalname)
});
const upload = multer({ storage: storage });

app.get('/health', (req, res) => {
    res.json({ status: "ok", node: NODE_NAME });
});

app.post('/upload/:filename', upload.single('file'), (req, res) => {
    res.json({ message: "File uploaded successfully" });
});

app.get('/download/:filename', (req, res) => {
    const filepath = path.join(STORAGE_DIR, req.params.filename);
    if (fs.existsSync(filepath)) {
        res.download(filepath);
    } else {
        res.status(404).json({ detail: "File not found on this node" });
    }
});

app.delete('/delete/:filename', (req, res) => {
    const filepath = path.join(STORAGE_DIR, req.params.filename);
    if (fs.existsSync(filepath)) {
        fs.unlinkSync(filepath);
        res.json({ message: "File deleted successfully" });
    } else {
        res.json({ message: "File not found or already deleted" });
    }
});

const __filename = fileURLToPath(import.meta.url);

if (process.argv[1] && __filename.includes(path.basename(process.argv[1]))) {
    const port = parseInt(process.argv[2] as string) || 8001;
    const nodeName = (process.argv[3] as string) || "node1";
    
    app.listen(port, "0.0.0.0", () => {
        console.log(`Node Server ${nodeName} running on port ${port}`);
    });
}

export default app;

File: api/node_server.py
from fastapi import FastAPI, UploadFile, File, HTTPException
from fastapi.responses import FileResponse
import os
import argparse

app = FastAPI()

NODE_NAME = os.getenv("NODE_NAME", "node1")
STORAGE_DIR = os.path.join("storage", NODE_NAME)

os.makedirs(STORAGE_DIR, exist_ok=True)

@app.get("/health")
def health_check():
    return {"status": "ok", "node": NODE_NAME}

@app.post("/upload/{filename}")
async def upload_file(filename: str, file: UploadFile = File(...)):
    filepath = os.path.join(STORAGE_DIR, filename)
    try:
        content = await file.read()
        with open(filepath, "wb") as f:
            f.write(content)
        return {"message": "File uploaded successfully"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@app.get("/download/{filename}")
def download_file(filename: str):
    filepath = os.path.join(STORAGE_DIR, filename)
    if os.path.exists(filepath):
        return FileResponse(filepath, filename=filename)
    raise HTTPException(status_code=404, detail="File not found on this node")

@app.delete("/delete/{filename}")
def delete_file(filename: str):
    filepath = os.path.join(STORAGE_DIR, filename)
    if os.path.exists(filepath):
        os.remove(filepath)
        return {"message": "File deleted successfully"}
    # Return 200 even if it doesn't exist to ensure idempotency across nodes
    return {"message": "File not found or already deleted"}

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--port", type=int, default=8001)
    parser.add_argument("--node", type=str, default="node1")
    args = parser.parse_args()
    
    os.environ["NODE_NAME"] = args.node
    NODE_NAME = args.node
    STORAGE_DIR = os.path.join("storage", NODE_NAME)
    os.makedirs(STORAGE_DIR, exist_ok=True)
    
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=args.port)
